### PR TITLE
Add conf column to legends

### LINKED ISSUE #10877: Add configuration json column to Legend
It turns out it's getting a bit more complicated than was originally thought. 

The original spirit of this ticket is to have a separate place for the column name and the column that's being used for technical purposes. It's been agreed that the title of the legend should be held in the `title` field and to introduce a `conf` or `params` JSON that should contain technical details for the legend to work properly.

cc/ @nobuti

### PR DESCRIPTION
**Pull request by gfiorav**
**DON'T DEPLOY BEFORE https://github.com/CartoDB/cartodb/issues/10954**

Fixes #10877

The idea is to add a validated JSON column to legends called conf to store things related to how the legend works and that need to be persisted (column names, etc)

cc @nobuti

### CODE REVIEW

**Comment by gfiorav** [comment]
Please CR @javitonino

File: db/migrate/20161130102840_add_conf_to_legends.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Sequel.migration do
+  up do
+    add_column :legends, :conf, :json, null: false, default: '{}'
+  end
+
+  down do
+    drop_column :legends, :conf
+  end
+end
```
Comment (javitonino) [inline db/migrate/20161130102840_add_conf_to_legends.rb]: Better to use the new migration wrapper to avoid locking issues.

Also, it goes without saying, split it to a different PR before deploying.

File: app/models/carto/legend.rb [hunk 2]
```diff
@@ -32,6 +35,10 @@ def ensure_definition
       self.definition ||= Hash.new
     end
 
+    def ensure_conf
+      self.conf ||= Hash.new
```
Comment (javitonino) [inline app/models/carto/legend.rb]: Is this needed even when the column is `NOT NULL` and has a default?

File: spec/models/carto/legend_spec.rb [hunk 3]
```diff
@@ -113,6 +113,59 @@ module Carto
...
+          bad_conf = { not_column_name: 'not array' }
+          legend = Legend.new(layer_id: @layer.id,
+                              type: 'bubble',
+                              definition: {},
+                              conf: bad_conf)
+
+          legend.valid?
+
+          legend.errors[:conf].should_not be_empty
+          legend.errors[:conf].should_not include('could not be validated')
```
Comment (javitonino) [inline spec/models/carto/legend_spec.rb]: Should not include that error message? Why?

File: app/models/carto/legend.rb (hunk 2 above)
Comment (gfiorav) [inline app/models/carto/legend.rb]: The JSON validator will swallow a `nil` on validation before db has the opportunity to add the default!

File: db/migrate/20161130102840_add_conf_to_legends.rb (hunk 1 above)
Comment (gfiorav) [inline db/migrate/20161130102840_add_conf_to_legends.rb]: Yeah, totally forgot, sorry.

File: spec/models/carto/legend_spec.rb (hunk 3 above)
Comment (gfiorav) [inline spec/models/carto/legend_spec.rb]: Careless copy paste, but still passes since that error only appears when something prevents the validation from happening (malformed json or whatever)

**Comment by gfiorav** [comment]
Can I get a 👍 ?

File: db/migrate/20161125173248_simplify_states_table.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+Sequel.migration do
```
Comment (javitonino) [inline db/migrate/20161125173248_simplify_states_table.rb]: 😱

File: db/migrate/20161125174333_create_snapshots.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+Sequel.migration do
```
Comment (javitonino) [inline db/migrate/20161125174333_create_snapshots.rb]: 😱

**Comment by gfiorav** [comment]
Have a look now @javitonino

**Comment by ivanmalagon** [comment]
Brought back from `acceptance` to `development`.

We found a bug in #10481 Perhaps some work in backend is needed. Until that frontend task is not fully OK, this current one must wait to pass acceptance and then be deployed.

For further information, please refer to the comment https://github.com/CartoDB/cartodb/issues/10481#issuecomment-264900788

**Comment by ivanmalagon** [comment]
Deploying it to `ded05` to accept only this backend part.

If new backend code must be developed because of https://github.com/CartoDB/cartodb/issues/10481#issuecomment-264900788 a new ticket should be created.

**Comment by javitonino** [comment]
Closing, merged via https://github.com/CartoDB/cartodb/pull/10980 due to Gemfile.lock conflicts
